### PR TITLE
[31] Link NIH clinical trials to OpenAlex papers

### LINKED ISSUE #31: [Translational]: Creating links between articles and trials

### PR DESCRIPTION
**Pull request by Jack-0-0**
Closes #31.

This PR adds:
* Pipeline to select trials with paper references and save as the intermediate dataset: `nih.data_processing.clinical_trials_with_references.intermediate`
* Pipeline to process the reference information into a format that can be used to link papers to clinical trials and save as the intermediate dataset: `nih.data_processing.clinical_trials_links_to_papers.intermediate`

### CODE REVIEW

File: conf/base/catalog.yml [hunk 1]
```diff
@@ -108,3 +112,11 @@ works_for_concepts:
 nih.data_collection.clinical_trials.raw:
   <<: *_csv_ptd
   path: s3://alphafold-impact/data/01_raw/nih/clinical_trials/
+
+nih.processing.clinical_trials_with_references.intermediate:
```
Comment (ampudia19) [inline conf/base/catalog.yml]: We probably want to keep naming conventions as standardised as possible, so I'd suggest changing these two:
```suggestion
nih.data_processing.clinical_trials_with_references.intermediate:
```

File: src/alphafold_impact/pipelines/data_processing_nih_clinical_trials/README.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# NIH Clinical Trials Data Processing Pipeline (`data_processing_nih_clinical_trials`)
+
+## Overview
+The `data_processing_nih_clinical_trials` pipeline processes raw data fetched from `clinicaltrials.gov`. This pipeline focuses on transforming the raw data into format suitable for further analysis within the project.
+
+## Modules
+### `nodes.py`
+This module contains the core functionalities for the pipeline:
+- `trials_with_references`: Loads NIH clinical trials from partitioned dataset, filters out trials that do not reference a paper, concatenates results into one DataFrame.
+- `trials_links_to_papers`: Produces a DataFrame which can be used to link NIH clinical trials to research papers via PubMed ID or DOI.
```
Comment (ampudia19) [inline src/alphafold_impact/pipelines/data_processing_nih_clinical_trials/README.md]: What wasn't obvious to me until I ran the code was that this has as input the output of `trials_with_references`. Maybe worth mentioning it for legacy sake.

File: src/alphafold_impact/pipelines/data_processing_nih_clinical_trials/__init__.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+"""
```
Comment (ampudia19) [inline src/alphafold_impact/pipelines/data_processing_nih_clinical_trials/__init__.py:1]: We should do something about the `__init__.py` docstrings :)

File: src/alphafold_impact/pipelines/data_processing_nih_clinical_trials/nodes.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        logger.info(
+            "Concatenated clinical trials with references from partition %s/%s",
+            idx,
+            n_partitions,
+        )
+
+    return ct_with_refs_combined
+
+
+def _reference_string_to_df(row: pd.Series) -> pd.DataFrame:
```
Comment (ampudia19) [inline src/alphafold_impact/pipelines/data_processing_nih_clinical_trials/nodes.py:24]: Neat! I wonder whether we should agree on how inner use functions are used - I may have a preference for listing them all at the top, but feel free to disregard.

File: src/alphafold_impact/pipelines/data_processing_nih_clinical_trials/nodes.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    clinical_trials_with_references: pd.DataFrame,
+) -> pd.DataFrame:
+    """
+    Processes clinical trial data with references to papers, to produce a DataFrame
+    which can be used to link NIH clinical trials to research papers via PubMed ID
+    or DOI.
+
+    This function takes a DataFrame containing clinical trial data with references
+    to papers and performs the following data processing steps:
+    1. Creates new columns for each component of the paper reference.
```
Comment (ampudia19) [inline src/alphafold_impact/pipelines/data_processing_nih_clinical_trials/nodes.py]: You may want to indent the list.

File: src/alphafold_impact/pipelines/data_processing_nih_clinical_trials/nodes.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        pd.DataFrame: A processed DataFrame containing the following columns:
+            - 'nct_id': The NCT ID of the clinical trial.
+            - 'ref_pmid': The PMID (PubMed ID) of the paper reference.
+            - 'ref_doi': The DOI (Digital Object Identifier) of the paper reference.
+            - 'ref_citation': The citation of the paper reference.
+            - 'ref_type': The type of the paper reference.
+            - 'ref_retraction_list': The retraction list associated with the paper reference.
+    """
+    return (
+        clinical_trials_with_references.apply(_reference_string_to_df, axis=1)
```
Comment (ampudia19) [inline src/alphafold_impact/pipelines/data_processing_nih_clinical_trials/nodes.py:128]: FYI this defeats a bit the purpose of pipelines, as this one currently only uses the parent function to process an entire pipeline under the hood that is not quite captured by `kedro viz`. Again, mostly a styling thing at this point, but for future pipelines that may require versioning and better separation of concerns, we may need to revisit pipelines :)

**Review by ampudia19 — APPROVED** [review]
Minor comments mostly around styling, with a single actual request (`processing` -> `data_processing`).

File: src/alphafold_impact/pipelines/data_processing_nih_clinical_trials/README.md (hunk 2 above)
Comment (Jack-0-0) [reply src/alphafold_impact/pipelines/data_processing_nih_clinical_trials/README.md]: I've added a note regarding this

File: src/alphafold_impact/pipelines/data_processing_nih_clinical_trials/__init__.py (hunk 3 above)
Comment (Jack-0-0) [reply src/alphafold_impact/pipelines/data_processing_nih_clinical_trials/__init__.py:1]: How about removing them all?

File: src/alphafold_impact/pipelines/data_processing_nih_clinical_trials/nodes.py (hunk 4 above)
Comment (Jack-0-0) [reply src/alphafold_impact/pipelines/data_processing_nih_clinical_trials/nodes.py:24]: I'm happy to have them at the top. I've moved them

**Comment by Jack-0-0** [comment]
Thanks for the review @ampudia19. I've made your suggested changes

**Pull request merged**